### PR TITLE
Replace cargo-make with regular Makefile

### LINKED ISSUE #1046: Investigate how to get rid of cargo-make
I propose to find a way for getting rid of cargo-make for our build setup. Here is why:

- **The Makefile is hard to understand because it is only the _diff_ to the default Makefile.** Hence, it is hard to reason about how this whole thing works. This makes it especially hard for new-comers to the project to get started.
- **It does not capture everything.** cargo-make does not take care of node, yarn and other dependencies. It is not a one-stop shop that takes care of everything.
- **It causes issues from time to time.** We have had several occasions where updates to cargo-make broke our build or where changes in the ecosystem (clippy stabilized) did not go very smooth.
- **The design is not very good.** This is more of a personal note but I just dislike the design and working with it. I find it cumbersome to achieve the tasks I want to do.

The TenX people use `nix` for their builds. There may also be other solutions.

This spike is about evaluating which tools may replace cargo-make to ease the build and setup process.

Cargo by itself actually supports custom commands in the form of aliases similarly to git: https://doc.rust-lang.org/cargo/reference/config.html#configuration-keys

### PR DESCRIPTION
**Pull request by thomaseizinger**
Closes #1046.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+TOOLCHAIN = $(shell cat rust-toolchain)
+CARGO = $(RUSTUP) run --install $(TOOLCHAIN) cargo --color always
+
+NIGHTLY_TOOLCHAIN = "nightly-2019-07-31"
+CARGO_NIGHTLY = $(RUSTUP) run --install $(NIGHTLY_TOOLCHAIN) cargo --color always
+
+GIT_HOOKS_PATH = ".githooks"
+GIT_HOOKS = $(wildcard $(GIT_HOOKS_PATH)/*)
+
+BUILD_ARGS = ""
```
Comment (D4nte) [inline Makefile]: I am not a makefile expert, can you test that indeed `-e` would override this?

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+install_rust_nightly:
+	$(RUSTUP) install $(NIGHTLY_TOOLCHAIN)
+
+## Dev environment
+
+install_clippy: install_rust
+	$(RUSTUP) component list --installed --toolchain $(TOOLCHAIN) | grep -q clippy || $(RUSTUP) component add clippy --toolchain $(TOOLCHAIN)
+
+# need nightly toolchain to get access to `merge_imports`
```
Comment (D4nte) [inline Makefile]: I think we use more than `merge_imports` on fmt nightly.. no? I'd just remove this comment.

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+install_rustfmt: install_rust_nightly
+	$(RUSTUP) component list --installed --toolchain $(NIGHTLY_TOOLCHAIN) | grep -q rustfmt || $(RUSTUP) component add rustfmt --toolchain $(NIGHTLY_TOOLCHAIN)
+
+install_tomlfmt: install_rust
+	$(CARGO) --list | grep -q tomlfmt || $(CARGO) install cargo-tomlfmt
+
+install_yarn:
```
Comment (D4nte) [inline Makefile]:
```suggestion
yarn_install:
```
Name is confusing as all `install_foo` install `foo` on the machine.
Here, we are running `yarn install`.

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+## Development tasks
+
+all: format build clippy test doc e2e_scripts
+
+format: install_rustfmt install_tomlfmt install_yarn
+	$(CARGO_NIGHTLY) fmt
+	$(CARGO) tomlfmt -p Cargo.toml
+	(cd ./api_tests; yarn run fix)
+
+ci: check_format clippy build test e2e
```
Comment (D4nte) [inline Makefile]: no `doc`?

File: Makefile (hunk 1 above)
Comment (thomaseizinger) [reply Makefile]: I tested it :)
`make` actually has a very useful `-n` option to dry-run a task!

File: Makefile (hunk 2 above)
Comment (thomaseizinger) [reply Makefile]: True!

**Comment by D4nte** [comment]
Also, maybe add BUILD_ARGS to test, clippy, doc etc tasks like I propose in my create-comit-app Release 0.5.0 PR

**Comment by thomaseizinger** [comment]
> Also, maybe add BUILD_ARGS to test, clippy, doc etc tasks like I propose in my create-comit-app Release 0.5.0 PR

Well those are not `build` commands. I can add `TEST_ARGS` etc or generalize it to `CARGO_ARGS` but I feel like that doesn't make any sense because you don't want to run the unit tests against `--release` I believe (is that even possible?)

**Comment by D4nte** [comment]
> > Also, maybe add BUILD_ARGS to test, clippy, doc etc tasks like I propose in my create-comit-app Release 0.5.0 PR
> 
> Well those are not `build` commands. I can add `TEST_ARGS` etc or generalize it to `CARGO_ARGS` but I feel like that doesn't make any sense because you don't want to run the unit tests against `--release` I believe (is that even possible?)

Leave it like this. I don't care about `--release` but more for other argument (features?) that one might want to pass down.

**Comment by thomaseizinger** [comment]
> > > Also, maybe add BUILD_ARGS to test, clippy, doc etc tasks like I propose in my create-comit-app Release 0.5.0 PR
> > 
> > 
> > Well those are not `build` commands. I can add `TEST_ARGS` etc or generalize it to `CARGO_ARGS` but I feel like that doesn't make any sense because you don't want to run the unit tests against `--release` I believe (is that even possible?)
> 
> Leave it like this. I don't care about `--release` but more for other argument (features?) that one might want to pass down.

Yeah good point. I was thinking about that but ended up not doing it because there is currently no need for it. I'd suggest we do very specific arguments for now (`BUILD_ARGS`, `INSTALL_ARGS`) depending on what we need. Once that gets annoying, we can always introduce something better :)

**Pull request merged**